subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/NonHierarchicalFileWatcherUpdater.java

```diff
@@ -158,11 +171,16 @@ private void updateWatchedDirectories(Map<String, Integer> changedWatchDirectori
     }
 
     private static void decrement(String path, Map<String, Integer> changedWatchedDirectories) {
-        changedWatchedDirectories.compute(path, (key, value) -> value == null ? -1 : value - 1);
+        changedWatchedDirectories.compute(path, (key, value) -> zeroToNull(value == null ? -1 : value - 1));
     }
 
     private static void increment(String path, Map<String, Integer> changedWatchedDirectories) {
-        changedWatchedDirectories.compute(path, (key, value) -> value == null ? 1 : value + 1);
+        changedWatchedDirectories.compute(path, (key, value) -> zeroToNull(value == null ? 1 : value + 1));
+    }
+
+    @Nullable
+    private static Integer zeroToNull(int value) {
+        return value == 0 ? null : value;
```

Comment:
Why not add a `nullToZero()` method as well? That'd make the code more readable I think.